clang/lib/Sema/CheckExprLifetime.cpp

```diff
@@ -1094,6 +1094,24 @@ static bool pathOnlyHandlesGslPointer(IndirectLocalPath &Path) {
   return false;
 }
 
+static bool
+isLifetimeboundInterleaveInGSL(llvm::ArrayRef<IndirectLocalPathEntry> PathRef) {
```

Comment:
I am just wondering if this is the right approach to filter this problem out.

I think in the problematic code example `[[clang::lifetimebound]]` describes the lifetime of the returned reference, not the lifetime of the `gsl::pointer` annotated type. So, I think the mismatch is trying to apply the lifetimebound logic to the `gsl::pointer` in this particular example.

I think my point is, probably not all interleavings are bad, but here the `gsl::pointer` and the `lifetimebound` apply to different levels of indirection and I am wondering it that needs to be part of our filtering.

Reply:
> I am just wondering if this is the right approach to filter this problem out.
> 
> I think in the problematic code example `[[clang::lifetimebound]]` describes the lifetime of the returned reference, not the lifetime of the `gsl::pointer` annotated type. So, I think the mismatch is trying to apply the lifetimebound logic to the `gsl::pointer` in this particular example.

From an implementation perspective, we use the `lifetimebound` attribute to decide whether to continue traversing inner AST nodes (the return type doesn’t matter; it only affects the emitted diagnostic message). This process of each step is stateless -- it does not account for whether the existing path has been initialized for GSL.

Back to the problematic case,

```
class [[gsl::Pointer]] FooPointer {};

template<typename T>
class [[gsl::Owner]] Container {
public:
  const T& get() const [[clang::lifetimebound]];
};

FooPointer abc2() {
  Container<FooPointer> foo;
  return foo.get(); // Clang emits a return-stack-address warning on `foo`.
}
```
When we report the local `foo` entity in the TemporaryVisitor callback:

- The Path is [GslPointerInit, Lifetimebound]
- foo is of type `Container<FooPointer>`

The code incorrectly interprets that the `FooPointer` is constructed from `foo`, which is causing the issue here. I think it is essentially a bug in the code that analyzes the gsl::pointer from gsl::owner construction case (addressing this in the existing gsl::pointer filtering logic seems like a natural approach, though it introduces new false negatives).

> I think my point is, probably not all interleavings are bad, but here the `gsl::pointer` and the `lifetimebound` apply to different levels of indirection and I am wondering it that needs to be part of our filtering.

I completely agree that not all interleavings are bad, and there are some useful examples of where they work well. Better ideas are welcome.

Reply:
> The Path is [GslPointerInit, Lifetimebound]

I was wondering if we can also look at the types here. I think initializing a `gsl::Pointer` with another lifetimebound `gsl::Pointer` is probably fine. But initializing a `gsl::Pointer` with a lifetimebound reference is not OK, because it is the reference that has the lifetime information but the `gsl::Pointer` is more likely to be initialized with the pointee of the reference rather than the reference itself in this particular scenario.

I am not opposed to the current approach though if what I suggest is hard to implement.

Reply:
> I was wondering if we can also look at the types here. 

I had a similar thought and did a quick attempt yesterday. While it resolved this specific case, it introduced a few new issues. I may need to dig a bit deeper into it, so I’ll explore this approach further.